File: daily/2025/3531_Count_Covered_Buildings.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int countCoveredBuildings(int n, vector<vector<int>>& buildings) {
        vector<int> x_largest(n+1, 0);
        vector<int> x_smallest(n+1, n+1);
        vector<int> y_largest(n+1, 0);
        vector<int> y_smallest(n+1, n+1);
        for (auto& building : buildings){
            int x = building[0];
            int y = building[1];
            x_largest[x] = max(x_largest[x], y);
            y_largest[y] = max(y_largest[y], x);
            x_smallest[x] = min(x_smallest[x], y);
            y_smallest[y] = min(y_smallest[y], x);
        }
        int ans = 0;
        for (auto& building : buildings){
            int x = building[0];
            int y = building[1];
            if (x_largest[x] > y && x_smallest[x] < y && y_largest[y] > x && y_smallest[y] < x) ans++;
        }
        return ans;
    }
};

int main(){
    int n = 3;
    vector<vector<int>> buildings = {{1,2},{2,2},{3,2},{2,1},{2,3}};
    // int n = 3;
    // vector<vector<int>> buildings = {{1,1},{1,2},{2,1},{2,2}};
    // int n = 5;
    // vector<vector<int>> buildings = {{1,3},{3,2},{3,3},{3,5},{5,3}};
    Solution S;
    cout << S.countCoveredBuildings(n, buildings) << endl;
    return 0;
}

File: daily_python/3531_Count_Covered_Buildings.py
import os
import sys
from typing import List

class Solution:
    def countCoveredBuildings(self, n: int, buildings: List[List[int]]) -> int:
        x_largest = [0] * (n + 1)
        y_largest = [0] * (n + 1)
        x_smallest = [n + 1] * (n + 1)
        y_smallest = [n + 1] * (n + 1)
        for [x, y] in buildings:
            x_largest[x] = max(x_largest[x], y)
            y_largest[y] = max(y_largest[y], x)
            x_smallest[x] = min(x_smallest[x], y)
            y_smallest[y] = min(y_smallest[y], x)
        ans = 0
        for [x, y] in buildings:
            if x_smallest[x] < y < x_largest[x] and y_smallest[y] < x < y_largest[y]:
                ans += 1
        return ans

def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            n = int(lines[0].split('=')[1].strip()[:-1])
            buildings = [list(map(int, building.split(','))) for building in lines[1].split('=')[1].strip()[2:-3].split('},{')]
            dataset.append((n, buildings))
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for n, buildings in dataset:
        results.append(solution.countCoveredBuildings(n, buildings))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')